File: 235_lowest_common_ancestor_of_a_binary_search_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        while (root != null) {
            int val = root.val;

            if (p.val > val && q.val > val) {
                root = root.right;
            } else if (p.val < val && q.val < val) {
                root = root.left;
            } else {
                return root;
            }
        }
        return null;
    }
}

File: python/235_lowest_common_ancestor_of_a_binary_search_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if p.val > q.val:
            p, q = q, p
        def recursive(node, p, q):
            if node.val >= p.val and node.val <= q.val:
                return node
            if node.val > p.val and node.val > q.val:
                return recursive(node.left, p, q)
            else:
                return recursive(node.right, p, q)

        return recursive(root, p, q)
